torchrl/data/replay_buffers/samplers.py

```diff
@@ -414,58 +417,57 @@ def sample(self, storage: Storage, batch_size: int) -> torch.Tensor:
-    def _add_or_extend(self, index: Union[int, torch.Tensor]) -> None:
-        priority = self.default_priority
-
-        if not (
-            isinstance(priority, float)
-            or len(priority) == 1
-            or len(priority) == len(index)
-        ):
-            raise RuntimeError(
-                "priority should be a scalar or an iterable of the same "
-                "length as index"
-            )
-        # make sure everything is cast to cpu
-        index = torch.as_tensor(index, device=torch.device("cpu"), dtype=torch.long)
-        priority = torch.as_tensor(priority, device=torch.device("cpu"))
-        # MaxValueWriter will set -1 for items in the data that we don't want
-        # to update. We therefore have to keep only the non-negative indices.
-        valid_index = index >= 0
-        if not valid_index.all():
-            if valid_index.any():
-                index = index[valid_index]
-                if priority.numel() > 1:
-                    priority = priority[valid_index]
-            else:
-                return
-
-        self._sum_tree[index] = priority
-        self._min_tree[index] = priority
+        # TODO: This is an option if we want to sample when the trees are not initialized, though it screws up PrioritizedSliceSampler
+        # if p_sum <= 0:
+        #     index, *_ = RandomSampler.sample(self, storage, batch_size)
+        #     device = index[0].device if isinstance(index, tuple) else index.device
+        #     weight = torch.ones(batch_size, device=device)
+        # else:
+        #     if p_min <= 0:
+        #         p_min = 1
+        #
+        #     # For some undefined reason, only np.random works here.
+        #     # All PT attempts fail, even when subsequently transformed into numpy
+        #     mass = np.random.uniform(0.0, p_sum, size=batch_size)
+        #     # mass = torch.zeros(batch_size, dtype=torch.double).uniform_(0.0, p_sum)
+        #     # mass = torch.rand(batch_size).mul_(p_sum)
+        #     index = self._sum_tree.scan_lower_bound(mass)
+        #     index = torch.as_tensor(index)
+        #     if not index.ndim:
+        #         index = index.unsqueeze(0)
+        #     index.clamp_max_(len(storage) - 1)
+        #     weight = torch.as_tensor(self._sum_tree[index])
+        #
+        #     # Importance sampling weight formula:
+        #     #   w_i = (p_i / sum(p) * N) ^ (-beta)
+        #     #   weight_i = w_i / max(w)
+        #     #   weight_i = (p_i / sum(p) * N) ^ (-beta) /
+        #     #       ((min(p) / sum(p) * N) ^ (-beta))
+        #     #   weight_i = ((p_i / sum(p) * N) / (min(p) / sum(p) * N)) ^ (-beta)
+        #     #   weight_i = (p_i / min(p)) ^ (-beta)
+        #     # weight = np.power(weight / (p_min + self._eps), -self._beta)
+        #     weight = torch.pow(weight / p_min, -self._beta)
+        #     if storage.ndim > 1:
+        #         index = torch.unravel_index(index, storage._storage.shape)
```

Comment:
@wertyuilife2 Happy to hear your thoughts about this

Basically the idea would be to sample uniformly if a priority hasn't been passed. I won't make it part of this PR but we could consider this in the future.

Reply:
@vmoens  I believe we should ensure that **priorities are initialized at all times** and throw an error when `p_sum <= 0` or `p_min <= 0`. We should design around `default_priority` to handle potential issues rather than ignoring them.

Even considering buffer silence bug detection, I don't recommend such an implementation. 

For example, if some user's incorrect behavior or a bug causes `len(storage)` to be larger than it should be, this implementation could result in undetected silent bugs.